orttraining/orttraining/core/session/training_session.cc

```diff
@@ -258,29 +251,58 @@ Status TrainingSession::ConfigureForTraining(
   return Status::OK();
 }
 
-static Status AddLossFunctionInternal(Graph& graph,
-                                      ILossFunction& loss_graph_builder,
-                                      const LossFunctionInfo& loss_func_info,
-                                      const std::string& loss_scale_input_name,
-                                      std::string& actual_loss_name) {
-  auto loss_function_graph_defs = loss_graph_builder(graph, loss_func_info);
-
-  if (!loss_scale_input_name.empty()) {
-    // add node to scale by loss_scale_input_name
-    TypeProto* loss_type_proto = loss_function_graph_defs.CreateTypeProto({1}, ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
-    actual_loss_name = graph.GenerateNodeArgName("scaled_loss");
-    loss_function_graph_defs.AddNodeDefs(
-        {NodeDef{
-            "Mul",
-            {ArgDef{loss_func_info.loss_name}, ArgDef{loss_scale_input_name, loss_type_proto}},
-            {ArgDef{actual_loss_name, loss_type_proto}},
-            NodeAttributes(),
-            actual_loss_name}});
+static Status AddLossScaling(
+    const std::string& loss_name,
+    Graph& graph, std::string* loss_scale_input_name, std::string& scaled_loss_name) {
+  if (!loss_scale_input_name) {
+    scaled_loss_name = loss_name;
+    return Status::OK();
+  }
+
+  // add node to scale loss_name by loss_scale_input_name
+  GraphAugmenter::GraphDefs defs{};
+  *loss_scale_input_name = graph.GenerateNodeArgName("loss_scale");
```

Comment:
Do we need to `delete loss_scale_input_name` before allocating a new string to it?

Reply:
this just assigns to the pointed-to string (checked to be not null earlier). a new string is not dynamically allocated here.